Add ETCI2021 Dataset

isaaccorley opened the pull request:
- Adds `torchgeo.datasets.ETCI2021` from the [ETCI 2021 Flood Detection Challenge](https://nasa-impact.github.io/etci2021/)
- Updated docs
- Added `tests/datasets/test_etci2021.py`
- Added sample data for tests under `tests/data/etci2021/`

Questions:
- If the dataset splits (train/val/test) can be downloaded as separate files from google drive (train.zip, val.zip, test.zip), should we only download the file for the desired split or should we download the entire dataset?
- The dataset has 2 masks (water body and flood). For the train/val splits I load them into a mask of shape `(2, h, w)`. However, the test set flood masks were never released so the water mask is loaded `(h, w)`. Does this seem right or should I load add the additional dimension `(1, h, w)` for consistency?

Closes #118

adamjstewart commented:
> should we only download the file for the desired split or should we download the entire dataset?

Just that split

> Does this seem right or should I load add the additional dimension (1, h, w) for consistency?

Prob (1, h w)

isaaccorley commented:
I refactored the filenames, urls, md5s, and directories into a `metadata` attribute which can be indexed by the split. Let me know if you have a more standardized approach to this already. Also updated per your suggestions.

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,249 @@
+
+    'The authors would like to thank the NASA Earth Science Data Systems Program,
+    NASA Digital Transformation AI/ML thrust, and IEEE GRSS for organizing the ETCI competition'.
+    """  # noqa: E501
```
`# noqa: E501` should be removed and lines should be wrapped to 88 characters. We only ignore line length errors for URLs where it isn't possible to split the line in half.

isaaccorley replied:
Split lines in half

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,249 @@
+    NASA Digital Transformation AI/ML thrust, and IEEE GRSS for organizing the ETCI competition'.
+    """  # noqa: E501
+
+    splits = ["train", "val", "test"]
```
This is redundant since you can use `self.metadata.keys()` instead.

isaaccorley replied:
Fixed

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,249 @@
+            vhs = glob.glob(os.path.join(folder, "vh", "*.png"))
+            water_masks = glob.glob(os.path.join(folder, "water_body_label", "*.png"))
+
+            if split == "test":
+                flood_masks = [""] * len(water_masks)
```
It would be good to add a comment explaining why this hack is necessary

isaaccorley replied:
I can just remove the `flood_mask` key from the files instead if that's what you prefer? This is due to the test set not having any ground truth flood_mask.

adamjstewart replied:
That's prob simpler, we can just check if the key exists before loading it

isaaccorley replied:
Removed "flood_mask" from keys in files for test set.

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,249 @@
+        Returns:
+            True if the dataset directories and split files are found, else False
+        """
+        directory = self.metadata[self.split]["directory"]
+        dirpath = os.path.join(self.root, directory)
+        if not os.path.exists(dirpath):
+            return False
+        return True
```
It might be better to check the md5sum of the downloaded files instead of just checking whether a directory exists. There is an open issue on the best way to handle this, not sure what that is just yet though. We'll prob refactor this later, so no need to change it for now.

isaaccorley replied:
There are many images so wouldn't this create a large list of md5s we have to check? I agree with you though. We can discuss it further in the issue.

adamjstewart replied:
You could just validate the zip file itself. See #99 for my thoughts on an "ideal" behavior.

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,249 @@
+                )
+        return files
+
+    def _load_image(self, path: str) -> Tensor:
```
This seems to be a pretty common action for `VisionDataset`s, we should probably move this to the `VisionDataset` base class at some point, can be done in a future PR.

adamjstewart reviewed (commented):
Mostly looks good, just had a few minor comments.

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,253 @@
+    * 2 binary masks per image representing water body & flood, respectively
+    * 2 polarization band images (VV, VH) of 3 RGB channels per band
+    * 3 RGB channels per band generated by the Hybrid Pluggable
+    Processing Pipeline 'hyp3'
```
I think this needs to be indented a couple spaces, otherwise it will end the bulleted list, same below

isaaccorley replied:
Thanks for catching this. I was unsure how this would show up in the docs when rendered.

isaaccorley replied:
Fixed

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,253 @@
+    If you use this dataset in your research, please add the following to your
+    acknowledgements section:
+
+    'The authors would like to thank the NASA Earth Science Data Systems Program,
+    NASA Digital Transformation AI/ML thrust, and IEEE GRSS for organizing
+    the ETCI competition'.
```
```suggestion
        The authors would like to thank the NASA Earth Science Data Systems Program,
        NASA Digital Transformation AI/ML thrust, and IEEE GRSS for organizing
        the ETCI competition'.
```
Quote blocks can be created by indenting: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks

isaaccorley replied:
Fixed per suggestion

adamjstewart commented on torchgeo/datasets/etci2021.py:
```diff
@@ -0,0 +1,253 @@
+    2. flood/water
+
+    If you use this dataset in your research, please add the following to your
+    acknowledgements section:
```
```suggestion
    acknowledgements section::
```
I think this might fix the Sphinx issue.

adamjstewart commented:
Looks like there are conflicts preventing this from being rebased and merged. Personally I never use merge commits for this very reason. I always rebase manually to resolve merge conflicts. For now I'll squash and merge.
